test/integration/puppetlabs/services/jruby_pool_manager/jruby_pool_int_test.clj

```diff
@@ -289,67 +192,41 @@
-           (is (true? (borrow-until-desired-borrow-count
-                       pool-context
-                       10)))
+           (future (jruby-core/flush-pool! pool-context))
 
-           ;; at this point, we still have the main flush in progress, waiting for us
-           ;; to release the two instances from the old pool.  we should also have
-           ;; caused a flush of one of the two instances in the new pool, meaning that
-           ;; exactly one of the two in the new pool should have the ruby constant defined.
-           (is (true? (check-jrubies-for-constant-counts pool-context 1 1)))
+           ;; at this point, the flush is blocked, waiting for us
+           ;; to release the two instances we're holding before it can start
 
-           ;; now we'll set the ruby constants on both instances in the new pool
-           (is (true? (set-constants-and-verify pool-context 2)))
 
            ;; now we're going to return instance2 to the pool.  This should cause it
-           ;; to get flushed, but after that, the main pool flush operation should
-           ;; pull it out of the old pool and create a new instance in the new pool
-           ;; to replace it.  So we should end up with 3 instances in the new pool,
-           ;; two of which should have the ruby constants and one of which should not.
+           ;; to get flushed. The main pool flush operation is still blocked.
            (jruby-core/return-to-pool instance2
                                       :max-borrows-flush-while-pool-flush-in-progress-test
                                       [])
-           (is (true? (check-jrubies-for-constant-counts pool-context 2 1))))
-
-         ;; now we'll set the ruby constant on the 3 instances in the new pool
-         (is (true? (set-constants-and-verify pool-context 3)))
-
-         ;; The flush should still not have completed at this point because
-         ;; the last instance from the old pool is still being borrowed.
-         ;; Wait for some arbitrary (but short) period of time to see that
-         ;; that the agent is still busy - presumably handling the flush.
-         (is (false? (await-for 50 pool-agent)))
+           ;; Wait until instance2 is returned
+           (is (jruby-testutils/wait-for-instances pool 3) "Timed out waiting for instance2 to return to pool"))
 
          ;; and finally, we return the last instance from the old pool
          (jruby-core/return-to-pool instance1
                                     :max-borrows-flush-while-pool-flush-in-progress-test
                                     [])
 
          ;; wait until the flush is complete
-         (is (true? (jruby-testutils/timed-await (jruby-agents/get-pool-agent pool-context)))
-             (str "timed out waiting for the flush to complete, stack:\n"
-                  (get-all-stack-traces-as-str))))
+         (is (jruby-testutils/wait-for-instances pool 4) "Timed out waiting for the flush to finish"))
```

Comment:
Yeah you're right that's a race condition. I think you're right that if we wait for the future to finish, it will return after the jrubies have been borrowed, which will ensure that the `wait-for-instances` call won't accidentally return until the pool has actually been refilled